docs/user/library/jdbc/postgis.rst

```diff
@@ -77,3 +77,7 @@ An optimization on spatial queries can be applied, to avoid a bug with PostGIS q
 This optimization can be enabled using a system-wide default from the command line::
   
   java -Dorg.geotools.data.postgis.largeGeometriesOptimize=true
+
+By default JsonArrayContains will be delegated to @> operator. It is possible to instead delegate it to jsonb_path_exists, which was added in postgres version 12, by setting following property::
+
+  java -Dorg.geotools.data.postgis.useJsonPathExists=true
```

Comment:
The decision could be automatic. The PostgisDialect is responsible for creating the FilterToSQL, and it has a field that contains the PostgreSQL version. The field could theoretically be null, but in practice, it won't be by the time you try to use it for a filter to SQL in real usage.

Suggestion: in both the prepared and non prepared SQL dialects, pass down a flag to indicate whether to use the new syntax, and default to the old one if the field happens to be null.
In theory if missing you could grab a connection from the store, but don't do it: the filter to SQL creation might be called from a method that already has one, and grabbing 2+ connections to complete any operations is deadlock prone (every thread manages to get the first connection, the pool gets depleted, no thread can grab the second, deadlock happens).

Reply:
Updated the both PG dialects to check the version and pass it down to FilterToSQL.